File: cloud-collect-core/src/main/java/com/windea/demo/cloudcollect/core/domain/entity/CollectTag.java
package com.windea.demo.cloudcollect.core.domain.entity;

import com.windea.demo.cloudcollect.core.validation.annotation.UniqueCollectTag;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;

import javax.persistence.*;
import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.Size;
import java.io.Serializable;
import java.time.LocalDateTime;

/**
 * 收藏的标签。
 * <p>
 * 一个收藏可以带有多个标签。
 */
@Data
@NoArgsConstructor
@UniqueCollectTag
@Entity
public class CollectTag implements Serializable {
	private static final long serialVersionUID = 4229783797803970576L;

	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	private Long id;

	/**
	 * 所属用户。
	 */
	@ManyToOne(cascade = CascadeType.MERGE, fetch = FetchType.EAGER)
	private User user;

	/**
	 * 名字。
	 */
	@NotEmpty(message = "validation.CollectTag.name.NotEmpty")
	@Size(min = 1, max = 32, message = "validation.CollectTag.name.Size")
	@Column(nullable = false, length = 32)
	private String name;

	/**
	 * 概述。
	 */
	@NotEmpty(message = "validation.CollectTag.summary.NotEmpty")
	@Size(min = 1, max = 255, message = "validation.CollectTag.summary.Size")
	@Column(nullable = false)
	private String summary;

	/**
	 * 创建时间。
	 */
	@CreatedDate
	@Column
	private LocalDateTime createdTime;

	/**
	 * 最后更新时间。
	 */
	@LastModifiedDate
	@Column
	private LocalDateTime lastModifiedTime;
}

File: cloud-collect-core/src/main/kotlin/com/windea/demo/cloudcollect/core/domain/entity/CollectTag.kt
package com.windea.demo.cloudcollect.core.domain.entity

import com.windea.demo.cloudcollect.core.validation.annotation.*
import com.windea.demo.cloudcollect.core.validation.group.*
import io.swagger.annotations.*
import org.springframework.data.annotation.*
import org.springframework.data.jpa.domain.support.*
import java.io.*
import java.time.*
import javax.persistence.*
import javax.persistence.Id
import javax.validation.constraints.*

@ApiModel("收藏的标签。一个收藏可以带有多个标签。")
@Entity
@EntityListeners(AuditingEntityListener::class)
@UniqueCollectTag(groups = [Create::class])
data class CollectTag(
	@ApiModelProperty("编号。")
	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	var id: Long? = null,
	
	@ApiModelProperty("名字。")
	@field:NotEmpty(message = "{validation.CollectTag.name.NotEmpty}", groups = [Create::class, Modify::class])
	@field:Size(min = 1, max = 64, message = "{validation.CollectTag.name.Size}", groups = [Create::class, Modify::class])
	@Column(nullable = false, length = 64)
	var name: String,
	
	@ApiModelProperty("概述。")
	@field:NotEmpty(message = "{validation.CollectTag.summary.NotEmpty}", groups = [Create::class, Modify::class])
	@field:Size(min = 1, max = 255, message = "{validation.CollectTag.summary.Size}", groups = [Create::class, Modify::class])
	@Column(nullable = false)
	var summary: String = "",
	
	@ApiModelProperty("所属用户。")
	@ManyToOne(cascade = [CascadeType.MERGE], fetch = FetchType.EAGER, optional = false)
	var user: User
) : Serializable {
	@ApiModelProperty("创建时间。")
	@Column
	@CreatedDate
	var createdTime: LocalDateTime? = null
	
	@ApiModelProperty("最后更新时间。")
	@Column
	@LastModifiedDate
	var lastModifiedTime: LocalDateTime? = null
	
	@ApiModelProperty("收藏数量。")
	@Transient
	var collectCount: Long = 0
	
	
	override fun equals(other: Any?) = other === this || (other is CollectTag && other.id == id)
	
	override fun hashCode() = id.hashCode()
}
